### ISSUE TITLE
Kestrel randomly locked and stop responding.

### ISSUE THREAD

**Issue by ZOXEXIVO** [aspnet/KestrelHttpServer#1267]
I run ASP.NET Core 1.1 on Ubuntu 16.04 in Docker container (with Ubuntu image). 
This works fine, but 3 times in two weeks I catch error, that Kestrel stop responding. 
No errors in Exception filter, in logs - only last record with start Action Executing, but not completed.
Why problem in Kestrel ? - Because it locked and not processing new requests.
At locked state, CPU usage is 4-5% (per core). I wait 4 hours, but not result.

TOP command result:

```
PID    USER      PR    NI    VIRT       RES          SHR S       %CPU      %MEM      TIME+       COMMAND
10788   root      20   0     11.396g    185060       53704 S     4.7       2.3       14:17.56     dotnet
```

Restarting docker container is solve issue.
I can't attach and trace it, because all tools, that I use - works only on Windows.

**Comment by cesarbs** [comment]
Just to be clear, are you running everything (host and containers) on Linux?

Are you doing remote access from a Windows console window? It could be this if so: https://github.com/aspnet/KestrelHttpServer/issues/1259

**Comment by ZOXEXIVO** [comment]
Yes, it's all running on Ubuntu. 2-core server with no server CPU (G4400)
I know problem with console locking in Windows, with mouse e.t.c, but this is **not** same issue, because docker run in background and I not access it directly. 
Ketrel doing something, but i don't understand what.
CPU load 4.7% per core (at locked state) is equals to 3-5 fast browser requests to main page

**Comment by ZOXEXIVO** [comment]
UPD. Just now server blocked again. 2 days after restarted container

**Comment by davidfowl** [comment]
@ZOXEXIVO are you passing a cancellation token to any of your response writing API calls (if you have any)

**Comment by ZOXEXIVO** [comment]
@davidfowl  No. Only simple async requests to MongoDB (simple blog app)
Now, server works 7 days without blocking behind nginx proxy

**Comment by philjones88** [comment]
We're seeing similar. I can exec into the container and see the process still running. It's not responding to curl. It's sat at 4-6% CPU constantly.

I can 100% reproduce it. It happens randomly under heavy load from our protractor unit tests hitting our web app, so just takes anywhere from 0-20 minutes to trigger.

**Comment by jchannon** [comment]
I work with @philjones88 and to confirm this is a self-contained app running 1.0.1 on a CentOS 7 container.

**Comment by ZOXEXIVO** [comment]
After using NGINX before Kestrel, problem disappear (2 weeks uptime)

**Comment by euclid47** [comment]
@ZOXEXIVO we are experiencing the same symptoms in the same containerized environment.  Since deploying nginx before kestrel , have you had any locks?

**Comment by ZOXEXIVO** [comment]
@euclid47 behind nginx server, it works perfect! I
Here, https://github.com/aspnet/KestrelHttpServer/issues/1304 problem was in StaticFiles middleware.

**Comment by runxc1** [comment]
Any idea when #1304 will be pushed out?   I am running .net core 1.1 in an Azure web app and am similarly having the problem where it will crash and never start serving content again until the app is swapped or restarted

**Comment by muratg** [comment]
@ZOXEXIVO @runxc1 #1304 should be released in Feb.

**Comment by Eilon** [comment]
Closing as a dup of https://github.com/aspnet/KestrelHttpServer/issues/1304.

**Comment by tavisca-syedave** [comment]
Hi,
Do we have any plans to ship v1.1.1  as we are also getting the same error while deploying dot net core web app on Linux docker - ubntu ?

**Comment by Eilon** [comment]
@tavisca-syedave we plan to have the patch out hopefully later this month.

**Comment by cesarbs** [comment]
Kestrel 1.1.1 is now released. Is anyone still running into this issue?

**Comment by mxpv** [comment]
Hi,
@cesarbs I still have this issue on 1.1.1 from time to time. I'm running Docker containers (17.03.1-ce) on Ubuntu 16.04, MVC 1.1.1 behind Nginx. Also happens under heavy load.

It accepts few requests normally and than stops to respond.
nginx returns 504 Gateway Time-out.

```
info: Microsoft.AspNetCore.Server.Kestrel[14]
      Connection id "0HL43F9NVFNNJ" communication error.
Microsoft.AspNetCore.Server.Kestrel.Internal.Networking.UvException: Error -32 EPIPE broken pipe
...
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://HOST.net/8rr0  
info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
      Executing action method Controllers.FeedController.Feed (XXX) with arguments (8rr0) - ModelState is Valid
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://HOST.net/download/hn5a/PwDJILthcjU.mp4  
info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
      Executing action method Controllers.DownloadController.Download (XXX) with arguments (hn5a, PwDJILthcjU) - ModelState is Valid
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://HOST.net/yZa1  
info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
      Executing action method Controllers.FeedController.Feed (XXX) with arguments (yZa1) - ModelState is Valid
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://HOST.net/download/8e1J/S8aDtWdsxtg.mp4  
info: Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker[1]
      Executing action method Controllers.DownloadController.Download (XXX) with arguments (8e1J, S8aDtWdsxtg) - ModelState is Valid
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://HOST.net/hotb  

```

**Comment by pieceofsummer** [comment]
I think I've just got it with 1.1.2 using unix socket (behind nginx). Will enable verbose logging and come back if I can reproduce it.

**Comment by ilyabreev** [comment]
Just got it too with 1.1.2 begind nginx.

**Comment by jeff-pang** [comment]
I also just got it on 1.1.2. Nginx reverse proxy to Kestrel. In my case it was working in production for more more than 1 month then suddenly this it stopped working. Now it stopped working every hour

**Comment by ZOXEXIVO** [comment]
Best advice -> upgrade to 2.,0

**Comment by jeff-pang** [comment]
@ZOXEXIVO Did upgrading to v2 fixed your problem?

**Comment by ZOXEXIVO** [comment]
@jeff-pang Problem disappeared when i host kestrel behind nginx

**Comment by jeff-pang** [comment]
@ZOXEXIVO I am already hosting Kestrel behind Nginx, but I will give it a try anyway. Thanks

**Comment by jeff-pang** [comment]
Upgrading to 2.0 fixed the issue

**Comment by vinhhrv** [comment]
The same issue on 2.0.0, please check this

**Comment by cesarbs** [comment]
@vinhhrv Could you upload your dump somewhere and share it with us?

**Comment by vinhhrv** [comment]
@cesarbs
i will send my dump to your email is cesarbs@gmail.com ?
please confirm, (my email is vdvinh@gmail.com)

**Comment by cesarbs** [comment]
Sounds good, though depending on the size you might not be able to email it.

**Comment by vinhhrv** [comment]
@cesarbs, i sent to your email.
Thanks for support.

**Comment by cesarbs** [comment]
@vinhhrv About this:

> telnet port is oke, but curl http => hangs

Can you share more details about the specific request that causes it to hang? Or does it hang on any request?

**Comment by cesarbs** [comment]
@pieceofsummer @ilyabreev @jeff-pang Can you share more details about the issues you've ran into with Kestrel 1.1.2? In @vinhhrv's case, it appears he is running into the connection limit that we introduced in 2.0. But that limit isn't there in 1.1.2, so that can't be the cause of your issues.

Could any of you share a memory dump captured when Kestrel hangs?

**Comment by anthonywanted** [comment]
i got a same problem and i send you the dump zipped
please give me some advice
Thanks

**Comment by sanishx** [comment]
I'm also having the problem, on 2.0.0 it appears the whole dotnet process 'stops' rather than just kestrel as I have some background threads that stop doing anything also. The dotnet memory usage goes down bigly when this happens also

**Comment by hasanoruc** [comment]
Hello. Same problem. But my problem is a little bit different. So... I restart server every one or two week. It looks like server getting tired. When tired is not responding and then nginx getting 502 error. I am testing mobile app and getting and posting more datas and server get this error every 2 days. And i restart the server and then solve my problem for two days. I use ubuntu's service from /etc/systemd/system. I create webapi.service and webui.service for this dotnet crash problem. Dotnet starting after crash but not healty and crashing more and more times. What is my problem. Please solve my problem. Thanks.

@muratg
@davidfowl

**Comment by sanishx** [comment]
I am also using Ubuntu and running through the supervisor system.

**Comment by cesarbs** [comment]
@anthonywanted @sanishx @hasanoruc We've seen issues where the server seemingly hangs due to resource exhaustion. It is always the case that somewhere in the application a sync call is being made (e.g. `SomethingAsync().GetAwaiter().GetResult()`). This is unfortunately hidden in unexpected places sometimes, like some Azure SDKs. Have you checked if that could be the case?

**Comment by hasanoruc** [comment]
@cesarbs i will check.

and. more information.
my system did not show this error when it was 12gb ram, 6cores processor.
but my system is now 2cores processors and 4gbram. this error look when my system has been smaller. under the load(looking 45 persons per second) %2 cpu usage(2cores). not problem. but i test my app and I proccess more operation and dotnet is crashing. it seems like sockets. but i dont know.
sorry for my bad english.

**Comment by benaadams** [comment]
@hasanoruc do you use non-async (e.g. blocking) apis? May be some solutions in https://github.com/aspnet/KestrelHttpServer/issues/2104#issuecomment-337170742

With 2 cores the threadpool will have a lower minthreads value; so you may hit threadpool starvation (it will inject threads; but not as fast as the blocking apis are stopping them)

**Comment by hasanoruc** [comment]
@benaadams No. My all api methods are async.
Like this:


        [HttpPost("postxmesajsil")]
        [Produces("application/json")]
        public async Task<bool> PostXamarinMesajsil([FromHeader]string token, [FromBody]JObject obj)
        {
            string identity = (string)obj["userId"];

            await TokenControl(token, identity);
            
            int userId = (int)obj["userId"];
            int friendId = (int)obj["friendId"];

            return await messagesDal.sp_mesajsilxamarin(userId, friendId);
        }

**Comment by ChenWenBrian** [comment]
I'm also having the problem, I'm using dotnet core 2.0.  It appears the whole dotnet process 'stops' rather than just kestrel as I have some background threads that stop doing anything also.  This issue only happens when a reverse proxy server(HAProxy) is setting in front of kestrel and  concurrency is high(above 50).

**Comment by sanishmistry** [comment]
I have isolated the class method that runs a few moments before these 'stops' occur. I think it possibly may be in something below, where I decode some XML, then encode it in JSON and then decode the JSON to easiliy get an ExpandObject for some XML.

            var client = new HttpClient();

            var response = await client.GetStringAsync(options.FeedUrl);
            XDocument doc = XDocument.Parse(response); //or XDocument.Load(path)
            
            string jsonText = JsonConvert.SerializeXNode(doc);

            return  JsonConvert.DeserializeObject<ExpandoObject>(jsonText);

**Comment by Drawaes** [comment]
You are not disposing the http client for a start. Was there an xmldocument.load () in there somewhere?

**Comment by benaadams** [comment]
@sanishmistry resuse (e.g. `static`) HttpClient or Dispose it (e.g. `using`)? You may be hitting max outbound connections + keep alives on `client`?

**Comment by benaadams** [comment]
Also pump up your outbound connection limits?
```csharp
// Add at Main()
System.Net.ServicePointManager.DefaultConnectionLimit = 256; // Max concurrent outbound requests`
```

**Comment by sanishmistry** [comment]
i will wrap the block inn using statements and so if it fixes it, I think it that was likely the issue as it seemed to happen exactly 7 days after starting the process, and that block ran at a fixed time interval. I was not aware HttpClients needed to be exposed!

**Comment by benaadams** [comment]
>  I was not aware HttpClients needed to be disposed!

You can also just use a single static one and not Dispose it. However you can't change things like `DefaultHeaders` per call as that wouldn't be thread safe

**Comment by hasanoruc** [comment]
I solved my own problem for ubuntu 16.04.
I ran `top` command and then `shift+m` and then `shift+h`. And i saw many many dotnet actions loading the ram and it did not died after time by itself and it did not clear the ram by itself. And i looked to google and i add this line to crontab.(/etc/crontab):

> */5 *   * * *   root    killall --older-than 40m -s SIGKILL dotnet

What is this?
Every 5 minutes run this command. What command? killall dotnet actions older than 40 minutes.
And extra note: Add this extra line to crontab for every 5 minutes to clear all buffer/cache. And restart crontab for this: 

> */5 *   * * *   root    sync && echo 3 > /proc/sys/vm/drop_caches

`sudo /etc/init.d/cron reload`
`sudo /etc/init.d/cron restart`

**Comment by halter73** [comment]
@hasanoruc What's starting your dotnet processes? When should the dotnet processing be stopping?

**Comment by hasanoruc** [comment]
@halter73 
If every request to the server and every transaction in the database causes a dotnet process in the ram in my ubuntu server, and this does not disappear by itself. I need to clear these operations. I can clean them with crontab. When you delete this information and the cache on the server, the user is connected a bit late. But if it remains within the time specified in the crontab, it reads again from the cache and fast.

**Comment by niemyjski** [comment]
I'm also trying to track down this issue with two of our sites. It seems to happen when there is load on the site and it doesn't fix itself. We use a lot of async redis calls / http client calls via the nest client

**Comment by halter73** [comment]
@niemyjski It could be threadpool starvation. Have you collected a memory dump when Kestrel isn't responding?

**Comment by ilyabreev** [comment]
Same problem here. Randomly stops responding and consumes 1-2% CPU.

**Comment by halter73** [comment]
@ilyabreev Any details?

Information that might be useful:

1. OS version
2. Whether or not your running under IIS/ANCM or behind another reverse proxy
2. Frequency of occurence
3. Amount of load the server was under when it stopped responding
4. A core dump taken when the server is not responding
5. Anything out of the ordinary that could be related.

Also, I recommend opening a new issue. Even if the behavior described in this issue is similar to what you're experiencing, this issue describes a bug that was fixed prior to the release of 2.0.0.

**Comment by niemyjski** [comment]
@halter73 yes, I have some memory dumps.

**Comment by halter73** [comment]
@niemyjski Can you share these? You can email me with a link at the address listed in my github profile. Thanks.

**Comment by ilyabreev** [comment]
@halter73 
1. **Ubuntu 14.04.5 LTS** as host system. `dotnet` process runs inside docker container which based on **ubuntu:xenial**
2. Running behind `nginx` as reverse proxy.
3. 0-6 times in a day.
4. Don't know exactly. I don't think it's constantly too high. Maybe some kind of CPU spikes occured before locking.
5. Will try to email you link to memory dump.
6. Nothing special.

**Comment by mazong1123** [comment]
As this issue has been closed, do we have another new issue to track down the "dotnet process stops" problem? We're facing the same issue as the dotnet process stops unexpectedly.

**Comment by muratg** [comment]
@mazong1123 please feel free to file another bug to discuss the current issues.

**Comment by mshanak** [comment]
This[ blog post](https://medium.com/@mshanak/soved-dotnet-core-too-many-open-files-in-system-when-using-postgress-with-entity-framework-c6e30eeff6d1) could help  you increasing max open files and configure Nginx to close finished request connections

**Comment by hasanoruc** [comment]
I solved my problem under the load ram but i still getting 502 bad gateway issue. I am looking /var/log/nginx/error.log and it show "upstream prematurely closed connection while reading response header from upstream" error.
and i googled and change my some buffer settings and looks more fine now.
```
        proxy_buffering off;
        proxy_buffer_size 32k;
```

**Comment by hasanoruc** [comment]
I solved my problem under the load ram but i still getting 502 bad gateway issue. I am looking /var/log/nginx/error.log and it show "upstream prematurely closed connection while reading response header from upstream" error.
and i googled and change my some buffer settings and change the backend kestrel proxy 127.0.0.1:5000 to localhost:5000 and looks more fine now.

        location / {
        proxy_pass http://backend;
        proxy_buffering on;
        proxy_buffers 256 4k;
        proxy_buffer_size 8k;
        }
........
   upstream backend {
   server localhost:5000  
   }

**Comment by hasanoruc** [comment]
I guess i solved my own problem. A few days ago, i looked nginx error logs in /var/log/nginx/error.log. and saw error logs this: "upstream prematurely closed connection while reading response header from upstream". And i think about it and i looked my bad nginx config. And I changed the following config:
```
location / {
proxy_pass http://backend;  #throw error this line
}
location /api {
proxy_pass http://localhost:5021; #this is not throwing an error because config is ok.
}
upstream backend {
server 127.0.0.1:5020; # this throwing error. because i think coming request going to 2 port. and confusing.
server 127.0.0.1:19920;
}
```

my new config is:
```
location / {
proxy_pass http://localhost:5020;
}
```
And not using the upstream because not necessary my system.

Sorry for my bad english,
and
Sorry for my this stupid error.

**Comment by niemyjski** [comment]
@halter73 just a heads up I've been running 1.5 weeks in production without issues. I've seen high thread count and have been trying to get the profiler attached in azure to take a look (without success). I'm thinking the high thread count is just normal. I got a dump but all the theads were native and empty stack traces. Currently serving up 5+ billion requests a month across 4-5 large azure web app instances. Seeing 70-80% cpu and a couple hundred threads.  I was able to narrow down one of my issues to this: https://github.com/StackExchange/StackExchange.Redis/issues/759

**Comment by benaadams** [comment]
I've made a BlockingDetector package to output warnings to the logs when you block: 
https://github.com/benaadams/Ben.BlockingDetector

Only detects when blocking actually happens; so doesn't pick up calls that may block but don't or warn about coding practices that lead to blocking (or blocking that happens at OS rather than .NET level e.g. File.Read) - but it may help pick up things.

**Comment by Itoktsnhc** [comment]
Same problem here, randomly no response on my asp.net core project on Azure app service (IIS and ASP.NET Core Module). 
Another related issue: https://github.com/aspnet/Home/issues/1950

The typical metric of requests indicators as shown below, when no response error happens, the number of threads being used in the dotnet core will jump up to 400 or 500 rapidly.

![image](https://user-images.githubusercontent.com/11204672/36834116-ba748fbe-1d6c-11e8-87b2-eb32ff7a1d0a.png)

**Comment by shaykeren** [comment]
Same problem in my environment, ASP NET CORE 2.0 published to ubuntu 16.04.
Application deployed in AWS with the following instances:
instance #1 contains our API which hang from time to time
instance #2 contains our static html files, nginx which host the static files and forward the requests to API (instance #1) and other instances which serves our other microservices.
(API Process hangs every 2-4 days)

**I saw that this process has 650 dead threads, is it relevant?**

(lldb) sos Threads
ThreadCount:      677
UnstartedThread:  0
BackgroundThread: 25
PendingThread:    0
DeadThread:       650
Hosted Runtime:   no

Any Idea?

**Comment by runxc1** [comment]
If you update to 2.1 do you still run into the same problem?  I am just curious if the switch from libuv to sockets inside of Kestrel fixes this issue?

**Comment by shaykeren** [comment]
I'm familiar with your post and we are using a single httpclient.

**Comment by mshanak** [comment]
@shaykeren do use the HTTPclient to send requests periodically to an Url? if so, use cron job instead.

**Comment by euclid47** [comment]
@mshanak @shaykeren when we experienced this issue it happened with async calls to the user identity and/or redis for cached user data. The only solution was to bounce the application.

**Comment by ZOXEXIVO** [comment]
Catch it many times last week
NET Core 2.1, new sockets transport. Application stuck up

**Comment by halter73** [comment]
@ZOXEXIVO Are you running on 2.1.0 or one of the patches? It's possible you could be running into dotnet/corefx#29785 if you're still on 2.1.0.

**Comment by HassanHashemi** [comment]
Facing same issue on Centos
```
dotnet --version:
2.1.302
```
I have used @benaadams blocking detector and saw plenty of blocking call, however i do not think that it could be the actual issue because the same apps runs on windows for days without any problem.

any idea?

**Comment by kyuz0** [comment]
Sorry to re-open this. We are experiencing this issue in our application as well, we run on CentOS in docker and behind nginx. Our app has had this issue since with 2.0 and also with 2.1. Increasing the thread pool seemed to help, but sometimes we experience kestrel hanging with very little load while for the majority of time we have long up times with high load, which makes it difficult to debug.

Has there been any update on this or any additional troubleshooting advice? We are planning to take a full core dump of kesterel/dotnet the next time this happens and look at what's happening, we suspect it's got to do with some deadlocks but we can't really see how our code would result in deadlocks, so maybe we are triggering some bug at the framework level.

**Comment by halter73** [comment]
@kyuz0 If increasing the number of threadpool threads helped, that's a good sign you're experiencing threadpool starvation. This could be caused by a deadlock like you suspect, but there could be other causes too like blocking I/O.

To find the exact cause, collecting a core dump is your best bet. Unfortunately, analyzing linux core dumps isn't yet quite as easy as it is on Windows, but you can find [a guide here](https://github.com/dotnet/coreclr/blob/cc8de13a8853a76fb3eb99cc209a4986eb3b2585/Documentation/building/debugging-instructions.md#debugging-coreclr-on-linux). "clrstack -all" should show all the managed stacks and what they're blocked on.

**Comment by kyuz0** [comment]
@halter73 thanks for your reply. Is there any debugging that we can enable in Kestrel to get some logs/alerts when the threadpool is full? The difficulty at the moment is that when the application reaches this state, there's simply nothing happening in the logs.

Something else I've seen in our code is that we make raw TCP connections to some endpoints like this:

```
using(var client = new TcpClient())
            {
                    try
                    {
                        var result = client.BeginConnect(IPAddress.Parse(externalIp), port, null, null);
                        var success = result.AsyncWaitHandle.WaitOne(TimeSpan.FromMilliseconds(TimeoutReadyCheckMilliseconds));
                        if (!success)
                        {
                            return false;
                        }

                        client.EndConnect(result);
                    }
                    catch (SocketException e)
                    {
                        return false;
                    }
            }

```

The using statement should ensure that the TcpClient gets disposed of in all cases, it should not matter that we don't explicitly call EndConnect() in all cases... at least this was my understanding, but maybe this is what's causing issues as the socket will not receive an explicit EndConnect in case success is false or in case of exceptions.

**Comment by halter73** [comment]
Blocking on the IAsyncResult returned by TcpClient.BeginConnect is a possible candidate for the cause of your threadpool starvation. To fix this, you'll want to call the newer [TcpClient.ConnectAsync](https://docs.microsoft.com/es-es/dotnet/api/system.net.sockets.tcpclient.connectasync?view=netframework-4.7.2) method and await it in an async method while being careful to avoid calls to Task.Wait(), Task.Result or anything else in your codebase that could block a thread waiting on a Task. 

The BeginConnect/EndConnect method pair implement an older asynchronous API pattern called the [Asynchronous Programing Model or APM](https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/asynchronous-programming-model-apm). It's possible to use APM APIs without blocking, but it's much more difficult. You're almost always better off using a Task-returning API with async/await over using an APM API.

Also, did you leave some code out in the TcpClient initialization sample? The using block ends right after the call to TcpClient.EndConnect which means the connection has only just been established.

As for threadpool starvation events, you could look for the [ThreadPoolWorkerThreadAdjustmentAdjustment event](https://docs.microsoft.com/en-us/dotnet/framework/performance/thread-pool-etw-events#threadpoolworkerthreadadjustment) with "0x06 - Starvation" Reason [using perfcollect and perfview](https://github.com/dotnet/coreclr/blob/488d2a70ba59c58b26aa3f4f988a42f6b8927146/Documentation/project-docs/linux-performance-tracing.md). That being said, if this issues causes any health checks to fail, it's probably easier to have a watchdog process collect a core dump before killing the server process when the health check fails.
